internal/controller/children.go

```diff
@@ -262,10 +272,37 @@ func (c ChildResourceUpdates) IsRunning(pod *v1.Pod) bool {
 	stsReady := c.StatefulSet.Manifest.Status.ReadyReplicas == 1 && c.StatefulSet.Manifest.Status.Replicas == 1
 	podExists := pod != nil
 	podReady := podExists && podIsReady(pod)
-	return stsReady && podReady
+	if healthcheckURL == nil {
+		log.Error(fmt.Errorf("healtcheck url is nil"), "failed to do the healtcheck", "sessionName", c.StatefulSet.Manifest.GetName(), "sessionNamespace", c.StatefulSet.Manifest.GetNamespace())
+		return false
+	}
+	// NOTE: This could be achieved with a healthcheck in k8s, however when we have rstudio in a session
+	// then hitting the url for the session leads to redirect loop that fails the readiness check after 10
+	// redirects. K8s does not have a way to specify whether healthchecks should follow redirects - they
+	// just always do. So to avoid this problem we send a request here and apply the logic we want.
+	req, err := http.NewRequestWithContext(ctx, "GET", healthcheckURL.String(), nil)
+	if err != nil {
+		log.Error(err, "failed to create a new request for the healtcheck", "sessionName", c.StatefulSet.Manifest.GetName(), "sessionNamespace", c.StatefulSet.Manifest.GetNamespace(), "url", healthcheckURL.String())
+		return false
+	}
+	healthCheckOK := true
+	// NOTE: If we send only a single request then we can still get a 502 when the session loads
+	for i := 0; i < 5; i++ {
+		res, err := healtcheckClient.Do(req)
+		if err != nil {
+			log.Error(err, "the healtcheck response failed", "sessionName", c.StatefulSet.Manifest.GetName(), "sessionNamespace", c.StatefulSet.Manifest.GetNamespace(), "url", healthcheckURL.String())
+			return false
+		}
+		healthCheckOK = healthCheckOK && (res.StatusCode < 400)
+		if !healthCheckOK {
+			break
+		}
+		time.Sleep(time.Millisecond * 1000)
+	}
+	return stsReady && podReady && healthCheckOK
 }
 
-func (c ChildResourceUpdates) State(cr *amaltheadevv1alpha1.AmaltheaSession, pod *v1.Pod) (amaltheadevv1alpha1.State, string) {
+func (c ChildResourceUpdates) State(ctx context.Context, cr *amaltheadevv1alpha1.AmaltheaSession, pod *v1.Pod) (amaltheadevv1alpha1.State, string) {
```

Comment:
I tried a lot of different things and nothing worked:
- k8s readiness checks follow redirects and can fail (for rstudio for example)
- single http request still results in a lot of 50x errors when the sessions opens
- 500ms sleeps in a similar loop with http requests results in fewer 50x errors but they still occur

I think it should be fine. This adds some time to how long it takes before a session is fully running. And as for what is the impact on the operator performance I don't really know. 

If we see problems I can make it so that the retries on this are tracked on the CR or we can even simply do it in memory. Since we can only ever have a single operator active anyhow.

Reply:
Don't re-invent the wheel, use k8s probes with TCP or Http. If RStudio is broken, we should fix that and not battle-tested probe logic from k8s.

Reply:
K8s readiness probes (like all probes) support a minimum success count before flipping to ready.

Reply:
Which RStudio image are you testing with?

Reply:
All of them have this problem we have seen this before. They keep redirecting to some other path to login and then they redirect back to the probe causing a loop. I would have to find another URL to set the probe to. But I cannot know what this is.

Reply:
All I want is the probes to not redirect. But I cannot make the k8s probes not redirect.

Reply:
I wonder if I can just use a TCP probe on the port... I will check if that will work - I am pretty sure that will not follow redirects.

Reply:
To "fix" rstudio it would require submitting code to rstudio itself. There is not much we can do about this in our images.

Reply:
Maybe we should try the actual health check endpoint if it exists: https://docs.posit.co/ide/server-pro/auditing_and_monitoring/server_health_checks.html

Reply:
I did a bunch of tests on my train ride to Lausanne. The TCP checks work. I had to tweak the initial delay and period a bit but they do the trick.